src/neural/loader.cc

```diff
@@ -48,14 +48,97 @@ namespace lczero {
+  return addr[0] + 256 * addr[1] + 65536 * addr[2] + 16777216 * addr[3];
+}
+
+// Read a .zip file containing a network file (just stored, not compressed a
+// second time) that is appended at the end of the lc0 executable. Such a zip
+// file can be generated using, for example, "zip -0 zipfile.zip net.pb.gz",
+// "7z a -mx=0 zipfile.zip net.pb.gz" or the equivalent options of any other
+// compression utility.
+
+std::string DecompressEmbedded(std::string str) {
+  constexpr int MOD_GZIP_ZLIB_WINDOWSIZE = 15;
+  constexpr uint8_t eocd_sig[12] = {0x50, 0x4b, 5, 6, 0, 0, 0, 0, 1, 0, 1, 0};
+  constexpr uint8_t header_sig[4] = {0x50, 0x4b, 3, 4};
+
+  // Check if a zip file "end of central directory record" is there, 22 bytes
+  // before the file end.
+  uint8_t *eocd_addr =
+      reinterpret_cast<uint8_t *>(str.data()) + str.size() - 22;
+  if (memcmp(eocd_addr, eocd_sig, sizeof(eocd_sig)) != 0) {
+    throw Exception("No embeded file detected.");
+  }
+
+  // Find the start of the zip file by subtracting the "central directory" size
+  // and offset.
+  uint8_t *start_addr =
+      eocd_addr - read_le(eocd_addr + 12) - read_le(eocd_addr + 16);
+
+  // Check for a local file header.
+  if (memcmp(start_addr, header_sig, sizeof(header_sig)) != 0) {
+    throw Exception("No embeded file header detected.");
+  }
+
+  z_stream zs;
+  memset(&zs, 0, sizeof(zs));
+
+  if (inflateInit2(&zs, MOD_GZIP_ZLIB_WINDOWSIZE + 16) != Z_OK) {
+    throw Exception("inflateInit failed while decompressing.");
+  }
+
+  // Read the first file in the zip, should be a stored network file.
+  uint32_t offsets = read_le(start_addr + 26);
+  zs.next_in = reinterpret_cast<Bytef *>(start_addr) + 30 + (offsets >> 16) +
+               (offsets & 0xffff);
+  zs.avail_in = read_le(start_addr + 18);
+
+  std::string filename =
+      std::string(reinterpret_cast<char *>(start_addr) + 30, offsets & 0xffff);
+  CERR << "Loading embedded weights file: " << filename;
+
+  int ret;
+  char outbuffer[32768];
+  std::string outstring;
+
+  do {
+    zs.next_out = reinterpret_cast<Bytef *>(outbuffer);
+    zs.avail_out = sizeof(outbuffer);
+
+    ret = inflate(&zs, 0);
+
+    if (outstring.size() < zs.total_out) {
+      outstring.append(outbuffer, zs.total_out - outstring.size());
+    }
+
+  } while (ret == Z_OK);
+
+  inflateEnd(&zs);
+
+  if (ret != Z_STREAM_END) {
+    std::ostringstream oss;
+    oss << "Exception during zlib decompression: (" << ret << ") " << zs.msg;
+    throw Exception(oss.str());
+  }
+
+  return outstring;
+}
+
 std::string DecompressGzip(const std::string& filename) {
   const int kStartingSize = 8 * 1024 * 1024;  // 8M
   std::string buffer;
   buffer.resize(kStartingSize);
   int bytes_read = 0;
 
   // Read whole file into a buffer.
-  const gzFile file = gzopen(filename.c_str(), "rb");
+  gzFile file = gzopen(filename.c_str(), "rb");
+#ifdef _WIN32
+  if (!file && filename == CommandLine::BinaryName()) {
+    // Try again with a .exe suffix.
```

Comment:
Why is it needed? `CommandLine()::BinaryName()` doesn't return extension? :-\ It should.

Reply:
I found it returns the name as used on the command line. Since using the .exe extension is not mandatory, it wasn't returned I didn't write it - this may be shell dependent.

Reply:
Uh, that Windows..

Apparently in windows stdlib.h there is https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/get-pgmptr
Which returns proper path to the binary and doesn't require <windows.h> or using windows kernel functions.
There's also GetModuleFileName() which does require windows.h and is a standard windows function..
Anyway, I think it's a bug in `BinaryName()` and must be fixed there somehow rather than working around it at the caller site..

Reply:
I started doing something like that, but we don't need the file extension anywhere else. So I decided to avoid platform dependent code - the block here would be safe without the `#ifdef _WIN32`, as it will only trigger on windows.

Reply:
What if it's `.com` file? :)
I think it's still better to do that properly, this does look like a hack..